x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/test_config_context.tsx

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useCallback, useContext } from 'react';
+
+export interface TestConfig {
+  documents?: object[] | undefined;
+  verbose: boolean;
+}
+
+interface TestConfigContext {
+  testConfig: TestConfig;
+  setCurrentTestConfig: (config: TestConfig) => void;
+}
+
+const TEST_CONFIG_DEFAULT_VALUE = {
+  testConfig: {
+    verbose: false,
+  },
+  setCurrentTestConfig: () => {},
+};
+
+const TestConfigContext = React.createContext<TestConfigContext>(TEST_CONFIG_DEFAULT_VALUE);
+
+export const useTestConfigContext = () => {
+  const ctx = useContext(TestConfigContext);
+  if (!ctx) {
+    throw new Error(
+      '"useTestConfigContext" can only be called inside of TestConfigContext.Provider!'
+    );
+  }
+  return ctx;
+};
+
+export const TestConfigContextProvider = ({
```

Comment:
Really nice use of context here! I would recommend the following alterations to simplify the API slightly:

1. Move `useTestConfig` hook into the TestConfigProvider (they should never be used separately)
2. Suggestion: rename the `PipelineForm` component to `as PipelineFormUI` in provider

```diff
diff --git i/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form_provider.tsx w/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form_provider.tsx
index 9390c427ee..57abea2309 100644
--- i/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form_provider.tsx
+++ w/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form_provider.tsx
@@ -6,15 +6,13 @@
 
 import React from 'react';
 
-import { PipelineForm, PipelineFormProps } from './pipeline_form';
-import { TestConfigContextProvider, useTestConfig } from './test_config_context';
+import { PipelineForm as PipelineFormUI, PipelineFormProps } from './pipeline_form';
+import { TestConfigContextProvider } from './test_config_context';
 
 export const PipelineFormProvider: React.FunctionComponent<PipelineFormProps> = passThroughProps => {
-  const testConfigContextValue = useTestConfig();
-
   return (
-    <TestConfigContextProvider value={testConfigContextValue}>
-      <PipelineForm {...passThroughProps} />
+    <TestConfigContextProvider>
+      <PipelineFormUI {...passThroughProps} />
     </TestConfigContextProvider>
   );
 };
diff --git i/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/test_config_context.tsx w/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/test_config_context.tsx
index 0e1ea79013..6840ebef28 100644
--- i/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/test_config_context.tsx
+++ w/x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/test_config_context.tsx
@@ -35,17 +35,7 @@ export const useTestConfigContext = () => {
   return ctx;
 };
 
-export const TestConfigContextProvider = ({
-  children,
-  value,
-}: {
-  value: TestConfigContext;
-  children: React.ReactNode;
-}) => {
-  return <TestConfigContext.Provider value={value}> {children} </TestConfigContext.Provider>;
-};
-
-export const useTestConfig = () => {
+export const TestConfigContextProvider = ({ children }: { children: React.ReactNode }) => {
   const [testConfig, setTestConfig] = useState<TestConfig>({
     verbose: false,
   });
@@ -54,8 +44,14 @@ export const useTestConfig = () => {
     setTestConfig(currentTestConfig);
   }, []);
 
-  return {
-    testConfig,
-    setCurrentTestConfig,
-  };
+  return (
+    <TestConfigContext.Provider
+      value={{
+        testConfig,
+        setCurrentTestConfig,
+      }}
+    >
+      {children}
+    </TestConfigContext.Provider>
+  );
 };
```

Reply:
Great suggestions 👍. The latest commit addresses this.